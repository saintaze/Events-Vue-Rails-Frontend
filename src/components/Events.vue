<template>
  <div class="events" v-if="myEvents.length" >
    <h2 class="events-header">Events ({{myEvents.length}})</h2>
    <table class="events-table">
      <thead>
        <tr>
          <th>No</th>
          <th>Title</th>
          <th>Start Date</th>
          <th>End Date</th>
          <th>Lasts All Day</th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="(event, index) in myEvents" :key="event.id">
          <td>{{index + 1}}</td>
          <td>{{event.title}}</td>
          <td>{{event.start | dateFormat}}</td>
          <td>{{event.end | dateFormat}}</td>
          <td>{{event.allDay ? 'Yes' : 'No'}}</td>
        </tr>
      </tbody>
    </table>
  </div>
</template>

<script>
import moment from 'moment';

export default {
  name: 'events',
  computed: {
    myEvents(){
      return this.$store.getters.events;
    }
  },
  filters: {
    dateFormat(dateStr) {
      return moment(dateStr).format("ddd, MMM Do YYYY, h:mm a");
    }
  }
}
</script>

<style lang="scss">
@import url('https://fonts.googleapis.com/css?family=Source+Sans+Pro:400,700');

.events {
  font-family:'Source Sans Pro', sans-serif;

  &-header {
    margin: 32px auto;
  }
  
  &-table {
    width: 100%;
	  border-collapse: collapse; 
  
    tr:nth-of-type(odd) { 
      background-color: #eee; 
    }

    td, th { 
      text-align: center;
      padding: 10px;
    }

    th { 
      background: #111; 
      color: white; 
      font-weight: normal;
      font-size: 16px;
    }

    td {
      font-size: 14px;
    }
  }
}

</style>