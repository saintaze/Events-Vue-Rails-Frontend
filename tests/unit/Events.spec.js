import { shallowMount, createLocalVue } from '@vue/test-utils';
import Events from '@/components/Events.vue';
import Vuex from 'vuex';
import moment from 'moment';

let getters;
let store;
let wrapper;

const localVue = createLocalVue();
localVue.use(Vuex);

// create store helper
const createStore = (overrides={}) => {
  const defaultStoreConfig = {
    state: {},
    getters: {},
    mutations: {},
    actions: {}
  }
  return new Vuex.Store({...defaultStoreConfig, ...overrides});
}

// create wrapper helper
const createWrapper = (overrides={}) => {
  const defaultMountingOptions = {
    localVue,
    store: createStore()
  }
  return shallowMount(Events, {...defaultMountingOptions, ...overrides});
}

afterEach(() => { wrapper.destroy() });

//mock events data
const mockEvents = [
  {
    "id": 1,
    "title": "this is all we have ",
    "start": "2020-07-28",
    "end": "2020-07-29",
    "allDay": 1
  },
  {
    "id": 4,
    "title": "walking all night ",
    "start": "2020-08-25T07:00:00+05:00",
    "end": "2020-08-25T10:00:00+05:00",
    "allDay": 0
  },
]

describe('Events', () => {
 
  // describe block when events array is not empty
  describe('Events are present (events.length > 0)', ()=>{

    beforeEach(() => {
      getters = { events: () => mockEvents }
      store = createStore({ getters });
      wrapper = createWrapper({ store });
    });

    it('should render html correctly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should display .events element', () => {
      expect(wrapper.find('.events').exists()).toBe(true);
    });

    it('h2 elements should show total number of events', () => {
      expect(wrapper.find('h2').text()).toBe(`Events (${mockEvents.length})`);
    });

    it(`should contain ${mockEvents.length} tr elements`, () => {
      expect(wrapper.findAll('tr').length).toBe(mockEvents.length + 1); // +1 for the header row
    });

    it(`should show correct date format in start date and end date td elements`, () => {
      const allRowEls = wrapper.findAll('tr');
      const lastRowEl = allRowEls.at(allRowEls.length - 1);
      const lastStartDateEl = lastRowEl.findAll('td').at(2);
      const lastEndDateEl = lastRowEl.findAll('td').at(3);
      const dateFormat = 'ddd, MMM Do YYYY, h:mm a';
      const lastMockEvent = mockEvents[mockEvents.length - 1];

      expect(lastStartDateEl.text()).toBe(`${moment(lastMockEvent.start).format(dateFormat)}`);
      expect(lastEndDateEl.text()).toBe(`${moment(lastMockEvent.end).format(dateFormat)}`);
    });
  }) 

  // describe block when events array is empty
  describe('Events are empty (events.length === 0)', ()=>{

    beforeEach(() => {
      getters = { events: () => [] }
      store = createStore({ getters });
      wrapper = createWrapper({ store });
    });

    it('should render html correctly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not display .events element', () => {
      expect(wrapper.find('.events').exists()).toBe(false);
    });
  });
});