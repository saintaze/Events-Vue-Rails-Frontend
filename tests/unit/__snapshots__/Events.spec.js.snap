// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events Events are empty (events.length === 0) should render html correctly 1`] = `""`;

exports[`Events Events are present (events.length > 0) should render html correctly 1`] = `
<div class="events">
  <h2 class="events-header">Events (2)</h2>
  <table class="events-table">
    <thead>
      <tr>
        <th>Id</th>
        <th>Title</th>
        <th>Start Date</th>
        <th>End Date</th>
        <th>Lasts All Day</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>1</td>
        <td>this is all we have </td>
        <td>Tue, Jul 28th 2020, 12:00 am</td>
        <td>Wed, Jul 29th 2020, 12:00 am</td>
        <td>Yes</td>
      </tr>
      <tr>
        <td>4</td>
        <td>walking all night </td>
        <td>Tue, Aug 25th 2020, 7:00 am</td>
        <td>Tue, Aug 25th 2020, 10:00 am</td>
        <td>No</td>
      </tr>
    </tbody>
  </table>
</div>
`;
